File: src/main/java/com/dyonovan/neotech/client/modelfactory/models/ModelConnectedTextures.java
package com.dyonovan.neotech.client.modelfactory.models;

import com.dyonovan.neotech.collections.ConnectedTextures;
import com.dyonovan.neotech.collections.DummyState;
import com.dyonovan.neotech.common.blocks.connected.BlockConnectedTextures;
import net.minecraft.block.Block;
import net.minecraft.block.state.IBlockState;
import net.minecraft.client.renderer.block.model.*;
import net.minecraft.client.renderer.texture.TextureAtlasSprite;
import net.minecraft.client.resources.model.IBakedModel;
import net.minecraft.client.resources.model.ModelRotation;
import net.minecraft.item.ItemStack;
import net.minecraft.util.BlockPos;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumWorldBlockLayer;
import net.minecraft.world.IBlockAccess;
import net.minecraftforge.client.MinecraftForgeClient;
import net.minecraftforge.client.model.ISmartBlockModel;
import net.minecraftforge.client.model.ISmartItemModel;

import javax.vecmath.Vector3f;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

@SuppressWarnings( "deprecation" )
public class ModelConnectedTextures implements ISmartBlockModel, ISmartItemModel {
    protected ConnectedTextures textures;
    protected IBlockAccess world;
    protected BlockPos pos;
    protected BlockConnectedTextures block;
    protected static FaceBakery faceBakery = new FaceBakery();

    public ModelConnectedTextures(DummyState state) {
        textures = ((BlockConnectedTextures)state.block).getConnectedTextures();
        block = (BlockConnectedTextures) state.block;
        world = state.blockAccess;
        pos = state.pos;
    }

    public ModelConnectedTextures(BlockConnectedTextures block) {
        textures = block.getConnectedTextures();
    }

    @Override
    public List getFaceQuads(EnumFacing face) {
        return new ArrayList<>();
    }

    public void drawFace(boolean[] connections, List<BakedQuad> list, ModelRotation rot) {
        BlockPartFace face = new BlockPartFace(null, 0, "", new BlockFaceUV(new float[]{0.0F, 0.0F, 16.0F, 16.0F}, 0));
        boolean scale = true;

        list.add(faceBakery.makeBakedQuad(new Vector3f(0.0F, 0.0F, 16.0F), new Vector3f(8.0F, 8.0F, 15.999F), face, textures.getTextureForCorner(2, connections), EnumFacing.SOUTH, rot, null, scale, true));
        list.add(faceBakery.makeBakedQuad(new Vector3f(8.0F, 0.0F, 16.0F), new Vector3f(16.0F, 8.0F, 15.999F), face, textures.getTextureForCorner(3, connections), EnumFacing.SOUTH, rot, null, scale, true));
        list.add(faceBakery.makeBakedQuad(new Vector3f(0.0F, 8.0F, 16.0F), new Vector3f(8.0F, 16.0F, 15.999F), face, textures.getTextureForCorner(0, connections), EnumFacing.SOUTH, rot, null, scale, true));
        list.add(faceBakery.makeBakedQuad(new Vector3f(8.0F, 8.0F, 16.0F), new Vector3f(16.0F, 16.0F, 15.999F), face, textures.getTextureForCorner(1, connections), EnumFacing.SOUTH, rot, null, scale, true));

        list.add(faceBakery.makeBakedQuad(new Vector3f(0.0F, 0.0F, 16.0F), new Vector3f(8.0F, 8.0F, 15.999F), face, textures.getTextureForCorner(6, connections), EnumFacing.SOUTH, rot, null, scale, true));
        list.add(faceBakery.makeBakedQuad(new Vector3f(8.0F, 0.0F, 16.0F), new Vector3f(16.0F, 8.0F, 15.999F), face, textures.getTextureForCorner(7, connections), EnumFacing.SOUTH, rot, null, scale, true));
        list.add(faceBakery.makeBakedQuad(new Vector3f(0.0F, 8.0F, 16.0F), new Vector3f(8.0F, 16.0F, 15.999F), face, textures.getTextureForCorner(4, connections), EnumFacing.SOUTH, rot, null, scale, true));
        list.add(faceBakery.makeBakedQuad(new Vector3f(8.0F, 8.0F, 16.0F), new Vector3f(16.0F, 16.0F, 15.999F), face, textures.getTextureForCorner(5, connections), EnumFacing.SOUTH, rot, null, scale, true));

        if (block != null && block.isTranslucent()) {
            list.add(faceBakery.makeBakedQuad(new Vector3f(0.0F, 0.0F, 16.0F), new Vector3f(8.0F, 8.0F, 15.999F), face, textures.getTextureForCorner(2, connections), EnumFacing.NORTH, rot, null, scale, true));
            list.add(faceBakery.makeBakedQuad(new Vector3f(8.0F, 0.0F, 16.0F), new Vector3f(16.0F, 8.0F, 15.999F), face, textures.getTextureForCorner(3, connections), EnumFacing.NORTH, rot, null, scale, true));
            list.add(faceBakery.makeBakedQuad(new Vector3f(0.0F, 8.0F, 16.0F), new Vector3f(8.0F, 16.0F, 15.999F), face, textures.getTextureForCorner(0, connections), EnumFacing.NORTH, rot, null, scale, true));
            list.add(faceBakery.makeBakedQuad(new Vector3f(8.0F, 8.0F, 16.0F), new Vector3f(16.0F, 16.0F, 15.999F), face, textures.getTextureForCorner(1, connections), EnumFacing.NORTH, rot, null, scale, true));

            list.add(faceBakery.makeBakedQuad(new Vector3f(0.0F, 0.0F, 16.0F), new Vector3f(8.0F, 8.0F, 15.999F), face, textures.getTextureForCorner(6, connections), EnumFacing.NORTH, rot, null, scale, true));
            list.add(faceBakery.makeBakedQuad(new Vector3f(8.0F, 0.0F, 16.0F), new Vector3f(16.0F, 8.0F, 15.999F), face, textures.getTextureForCorner(7, connections), EnumFacing.NORTH, rot, null, scale, true));
            list.add(faceBakery.makeBakedQuad(new Vector3f(0.0F, 8.0F, 16.0F), new Vector3f(8.0F, 16.0F, 15.999F), face, textures.getTextureForCorner(4, connections), EnumFacing.NORTH, rot, null, scale, true));
            list.add(faceBakery.makeBakedQuad(new Vector3f(8.0F, 8.0F, 16.0F), new Vector3f(16.0F, 16.0F, 15.999F), face, textures.getTextureForCorner(5, connections), EnumFacing.NORTH, rot, null, scale, true));

        }
    }

    @Override
    public List getGeneralQuads() {
        ArrayList<BakedQuad> list = new ArrayList<>();
        boolean[] connections = new boolean[16];

        //Item (DON'T LOOK AROUND!)
        if (world == null) {
            Arrays.fill(connections, false);
            for (EnumFacing facing : EnumFacing.values())
                drawFace(connections, list, lookUpRotationForFace(facing));
            return list;
        }
        for(EnumFacing dir : EnumFacing.values()) {
            if (world.isAirBlock(pos.offset(dir)) || (!world.getBlockState(pos.offset(dir)).getBlock().isOpaqueCube() &&
                    !block.canBlockConnect(world.getBlockState(pos.offset(dir)).getBlock()))) {
                switch (dir) {
                    case UP:
                        connections[0] = block.canBlockConnect(world.getBlockState(pos.add(-1, 0, -1)).getBlock());
                        connections[1] = block.canBlockConnect(world.getBlockState(pos.add(0, 0, -1)).getBlock());
                        connections[2] = block.canBlockConnect(world.getBlockState(pos.add(1, 0, -1)).getBlock());
                        connections[3] = block.canBlockConnect(world.getBlockState(pos.add(-1, 0, 0)).getBlock());
                        connections[4] = block.canBlockConnect(world.getBlockState(pos.add(1, 0, 0)).getBlock());
                        connections[5] = block.canBlockConnect(world.getBlockState(pos.add(-1, 0, 1)).getBlock());
                        connections[6] = block.canBlockConnect(world.getBlockState(pos.add(0, 0, 1)).getBlock());
                        connections[7] = block.canBlockConnect(world.getBlockState(pos.add(1, 0, 1)).getBlock());
                        connections[8] = block.canBlockConnect(world.getBlockState(pos.add(-1, 1, -1)).getBlock());
                        connections[9] = block.canBlockConnect(world.getBlockState(pos.add(0, 1, -1)).getBlock());
                        connections[10] = block.canBlockConnect(world.getBlockState(pos.add(1, 1, -1)).getBlock());
                        connections[11] = block.canBlockConnect(world.getBlockState(pos.add(-1, 1, 0)).getBlock());
                        connections[12] = block.canBlockConnect(world.getBlockState(pos.add(1, 1, 0)).getBlock());
                        connections[13] = block.canBlockConnect(world.getBlockState(pos.add(-1, 1, 1)).getBlock());
                        connections[14] = block.canBlockConnect(world.getBlockState(pos.add(0, 1, 1)).getBlock());
                        connections[15] = block.canBlockConnect(world.getBlockState(pos.add(1, 1, 1)).getBlock());
                        drawFace(connections, list, lookUpRotationForFace(dir));
                        break;
                    case DOWN:
                        connections[0] = block.canBlockConnect(world.getBlockState(pos.add(-1, 0, 1)).getBlock());
                        connections[1] = block.canBlockConnect(world.getBlockState(pos.add(0, 0, 1)).getBlock());
                        connections[2] = block.canBlockConnect(world.getBlockState(pos.add(1, 0, 1)).getBlock());
                        connections[3] = block.canBlockConnect(world.getBlockState(pos.add(-1, 0, 0)).getBlock());
                        connections[4] = block.canBlockConnect(world.getBlockState(pos.add(1, 0, 0)).getBlock());
                        connections[5] = block.canBlockConnect(world.getBlockState(pos.add(-1, 0, -1)).getBlock());
                        connections[6] = block.canBlockConnect(world.getBlockState(pos.add(0, 0, -1)).getBlock());
                        connections[7] = block.canBlockConnect(world.getBlockState(pos.add(1, 0, -1)).getBlock());
                        connections[8] = block.canBlockConnect(world.getBlockState(pos.add(-1, -1, 1)).getBlock());
                        connections[9] = block.canBlockConnect(world.getBlockState(pos.add(0, -1, 1)).getBlock());
                        connections[10] = block.canBlockConnect(world.getBlockState(pos.add(1, -1, 1)).getBlock());
                        connections[11] = block.canBlockConnect(world.getBlockState(pos.add(-1, -1, 0)).getBlock());
                        connections[12] = block.canBlockConnect(world.getBlockState(pos.add(1, -1, 0)).getBlock());
                        connections[13] = block.canBlockConnect(world.getBlockState(pos.add(-1, -1, -1)).getBlock());
                        connections[14] = block.canBlockConnect(world.getBlockState(pos.add(0, -1, -1)).getBlock());
                        connections[15] = block.canBlockConnect(world.getBlockState(pos.add(1, -1, -1)).getBlock());
                        drawFace(connections, list, lookUpRotationForFace(dir));
                        break;
                    case NORTH:
                        connections[0] = block.canBlockConnect(world.getBlockState(pos.add(1, 1, 0)).getBlock());
                        connections[1] = block.canBlockConnect(world.getBlockState(pos.add(0, 1, 0)).getBlock());
                        connections[2] = block.canBlockConnect(world.getBlockState(pos.add(-1, 1, 0)).getBlock());
                        connections[3] = block.canBlockConnect(world.getBlockState(pos.add(1, 0, 0)).getBlock());
                        connections[4] = block.canBlockConnect(world.getBlockState(pos.add(-1, 0, 0)).getBlock());
                        connections[5] = block.canBlockConnect(world.getBlockState(pos.add(1, -1, 0)).getBlock());
                        connections[6] = block.canBlockConnect(world.getBlockState(pos.add(0, -1, 0)).getBlock());
                        connections[7] = block.canBlockConnect(world.getBlockState(pos.add(-1, -1, 0)).getBlock());
                        connections[8] = block.canBlockConnect(world.getBlockState(pos.add(1, 1, -1)).getBlock());
                        connections[9] = block.canBlockConnect(world.getBlockState(pos.add(0, 1, -1)).getBlock());
                        connections[10] = block.canBlockConnect(world.getBlockState(pos.add(-1, 1, -1)).getBlock());
                        connections[11] = block.canBlockConnect(world.getBlockState(pos.add(1, 0, -1)).getBlock());
                        connections[12] = block.canBlockConnect(world.getBlockState(pos.add(-1, 0, -1)).getBlock());
                        connections[13] = block.canBlockConnect(world.getBlockState(pos.add(1, -1, -1)).getBlock());
                        connections[14] = block.canBlockConnect(world.getBlockState(pos.add(0, -1, -1)).getBlock());
                        connections[15] = block.canBlockConnect(world.getBlockState(pos.add(-1, -1, -1)).getBlock());
                        drawFace(connections, list, lookUpRotationForFace(dir));
                        break;
                    case SOUTH:
                        connections[0] = block.canBlockConnect(world.getBlockState(pos.add(-1, 1, 0)).getBlock());
                        connections[1] = block.canBlockConnect(world.getBlockState(pos.add(0, 1, 0)).getBlock());
                        connections[2] = block.canBlockConnect(world.getBlockState(pos.add(1, 1, 0)).getBlock());
                        connections[3] = block.canBlockConnect(world.getBlockState(pos.add(-1, 0, 0)).getBlock());
                        connections[4] = block.canBlockConnect(world.getBlockState(pos.add(1, 0, 0)).getBlock());
                        connections[5] = block.canBlockConnect(world.getBlockState(pos.add(-1, -1, 0)).getBlock());
                        connections[6] = block.canBlockConnect(world.getBlockState(pos.add(0, -1, 0)).getBlock());
                        connections[7] = block.canBlockConnect(world.getBlockState(pos.add(1, -1, 0)).getBlock());
                        connections[8] = block.canBlockConnect(world.getBlockState(pos.add(-1, 1, 1)).getBlock());
                        connections[9] = block.canBlockConnect(world.getBlockState(pos.add(0, 1, 1)).getBlock());
                        connections[10] = block.canBlockConnect(world.getBlockState(pos.add(1, 1, 1)).getBlock());
                        connections[11] = block.canBlockConnect(world.getBlockState(pos.add(-1, 0, 1)).getBlock());
                        connections[12] = block.canBlockConnect(world.getBlockState(pos.add(1, 0, 1)).getBlock());
                        connections[13] = block.canBlockConnect(world.getBlockState(pos.add(-1, -1, 1)).getBlock());
                        connections[14] = block.canBlockConnect(world.getBlockState(pos.add(0, -1, 1)).getBlock());
                        connections[15] = block.canBlockConnect(world.getBlockState(pos.add(1, -1, 1)).getBlock());
                        drawFace(connections, list, lookUpRotationForFace(dir));
                        break;
                    case WEST:
                        connections[0] = block.canBlockConnect(world.getBlockState(pos.add(0, 1, -1)).getBlock());
                        connections[1] = block.canBlockConnect(world.getBlockState(pos.add(0, 1, 0)).getBlock());
                        connections[2] = block.canBlockConnect(world.getBlockState(pos.add(0, 1, 1)).getBlock());
                        connections[3] = block.canBlockConnect(world.getBlockState(pos.add(0, 0, -1)).getBlock());
                        connections[4] = block.canBlockConnect(world.getBlockState(pos.add(0, 0, 1)).getBlock());
                        connections[5] = block.canBlockConnect(world.getBlockState(pos.add(0, -1, -1)).getBlock());
                        connections[6] = block.canBlockConnect(world.getBlockState(pos.add(0, -1, 0)).getBlock());
                        connections[7] = block.canBlockConnect(world.getBlockState(pos.add(0, -1, 1)).getBlock());
                        connections[8] = block.canBlockConnect(world.getBlockState(pos.add(-1, 1, -1)).getBlock());
                        connections[9] = block.canBlockConnect(world.getBlockState(pos.add(-1, 1, 0)).getBlock());
                        connections[10] = block.canBlockConnect(world.getBlockState(pos.add(-1, 1, 1)).getBlock());
                        connections[11] = block.canBlockConnect(world.getBlockState(pos.add(-1, 0, -1)).getBlock());
                        connections[12] = block.canBlockConnect(world.getBlockState(pos.add(-1, 0, 1)).getBlock());
                        connections[13] = block.canBlockConnect(world.getBlockState(pos.add(-1, -1, -1)).getBlock());
                        connections[14] = block.canBlockConnect(world.getBlockState(pos.add(-1, -1, 0)).getBlock());
                        connections[15] = block.canBlockConnect(world.getBlockState(pos.add(-1, -1, 1)).getBlock());
                        drawFace(connections, list, lookUpRotationForFace(dir));
                        break;
                    case EAST:
                        connections[0] = block.canBlockConnect(world.getBlockState(pos.add(0, 1, 1)).getBlock());
                        connections[1] = block.canBlockConnect(world.getBlockState(pos.add(0, 1, 0)).getBlock());
                        connections[2] = block.canBlockConnect(world.getBlockState(pos.add(0, 1, -1)).getBlock());
                        connections[3] = block.canBlockConnect(world.getBlockState(pos.add(0, 0, 1)).getBlock());
                        connections[4] = block.canBlockConnect(world.getBlockState(pos.add(0, 0, -1)).getBlock());
                        connections[5] = block.canBlockConnect(world.getBlockState(pos.add(0, -1, 1)).getBlock());
                        connections[6] = block.canBlockConnect(world.getBlockState(pos.add(0, -1, 0)).getBlock());
                        connections[7] = block.canBlockConnect(world.getBlockState(pos.add(0, -1, -1)).getBlock());
                        connections[8] = block.canBlockConnect(world.getBlockState(pos.add(1, 1, 1)).getBlock());
                        connections[9] = block.canBlockConnect(world.getBlockState(pos.add(1, 1, 0)).getBlock());
                        connections[10] = block.canBlockConnect(world.getBlockState(pos.add(1, 1, -1)).getBlock());
                        connections[11] = block.canBlockConnect(world.getBlockState(pos.add(1, 0, 1)).getBlock());
                        connections[12] = block.canBlockConnect(world.getBlockState(pos.add(1, 0, -1)).getBlock());
                        connections[13] = block.canBlockConnect(world.getBlockState(pos.add(1, -1, 1)).getBlock());
                        connections[14] = block.canBlockConnect(world.getBlockState(pos.add(1, -1, 0)).getBlock());
                        connections[15] = block.canBlockConnect(world.getBlockState(pos.add(1, -1, -1)).getBlock());
                        drawFace(connections, list, lookUpRotationForFace(dir));
                        break;
                    default:
                        break;
                }
            }
        }
        return list;
    }

    public ModelRotation lookUpRotationForFace(EnumFacing face) {
        switch (face) {
            case UP :
                return ModelRotation.X90_Y0;
            case DOWN:
                return ModelRotation.X270_Y0;
            case NORTH:
                return ModelRotation.X0_Y180;
            case EAST:
                return ModelRotation.X0_Y270;
            case SOUTH:
                return ModelRotation.X0_Y0;
            case WEST:
                return ModelRotation.X0_Y90;
            default:
                return ModelRotation.X0_Y0;
        }
    }

    @Override
    public boolean isAmbientOcclusion() {
        return true;
    }

    @Override
    public boolean isGui3d() {
        return true;
    }

    @Override
    public boolean isBuiltInRenderer() {
        return false;
    }

    @Override
    public IBakedModel handleBlockState(IBlockState state) {
        return state instanceof DummyState ? new ModelConnectedTextures((DummyState)state) : null;
    }

    @Override
    public IBakedModel handleItemState(ItemStack stack) {
        return Block.getBlockFromItem(stack.getItem()) instanceof BlockConnectedTextures ? new ModelConnectedTextures((BlockConnectedTextures) Block.getBlockFromItem(stack.getItem())) : null;
    }

    @Override
    public TextureAtlasSprite getTexture() {
        return textures.corners;
    }

    @Override
    public ItemCameraTransforms getItemCameraTransforms() {
        return new ItemCameraTransforms(ModelBlock.MovedUp, ItemTransformVec3f.DEFAULT, ItemTransformVec3f.DEFAULT, ItemTransformVec3f.DEFAULT);
    }
}

File: src/main/scala/com/dyonovan/neotech/client/modelfactory/models/ModelConnectedTextures.scala
package com.dyonovan.neotech.client.modelfactory.models

import java.util

import com.dyonovan.neotech.common.blocks.{BlockConnectedTextures, ConnectedTexturesState}
import net.minecraft.block.Block
import net.minecraft.block.state.IBlockState
import net.minecraft.client.renderer.block.model._
import net.minecraft.client.renderer.texture.TextureAtlasSprite
import net.minecraft.client.resources.model.{IBakedModel, ModelRotation}
import net.minecraft.item.ItemStack
import net.minecraft.util.{BlockPos, EnumFacing}
import net.minecraft.world.IBlockAccess
import net.minecraftforge.client.model.{TRSRTransformation, ISmartBlockModel, ISmartItemModel}
import org.lwjgl.util.vector.Vector3f

/**
  * This file was created for NeoTech
  *
  * NeoTech is licensed under the
  * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
  * http://creativecommons.org/licenses/by-nc-sa/4.0/
  *
  * @author Paul Davis "pauljoda"
  * @since 2/26/2016
  */
class ModelConnectedTextures extends ISmartBlockModel with ISmartItemModel {
    lazy val faceBakery = new FaceBakery
    var block : BlockConnectedTextures = null
    var world : IBlockAccess = null
    var pos : BlockPos = null

    override def handleBlockState(state: IBlockState): IBakedModel = {
        state match {
            case connectedState : ConnectedTexturesState =>
                world = connectedState.world
                pos = connectedState.pos
                block = connectedState.block
                this
            case _ => this
        }
    }

    override def handleItemState(stack: ItemStack): IBakedModel = {
        block = Block.getBlockFromItem(stack.getItem).asInstanceOf[BlockConnectedTextures]
        this
    }

    override def getFaceQuads(facing : EnumFacing): util.List[BakedQuad] = {
        val bakedQuads = new util.ArrayList[BakedQuad]()
        if(world != null && world.getBlockState(pos.offset(facing)).getBlock != block)
            drawFace(block.getConnectionArrayForFace(world, pos, facing), bakedQuads, lookUpRotationForFace(facing), facing.getOpposite)
        else if(world == null)
            drawFace(Array[Boolean](false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false), bakedQuads, lookUpRotationForFace(facing), facing)

        bakedQuads
    }

    def drawFace(connections : Array[Boolean], list : java.util.List[BakedQuad], rot : ModelRotation, facing : EnumFacing) {
        val face : BlockPartFace = new BlockPartFace(null, 0, "", new BlockFaceUV(Array[Float](0.0F, 0.0F, 16.0F, 16.0F), 0))
        val scale : Boolean = true

        list.add(faceBakery.makeBakedQuad(new Vector3f(0.0F, 0.0F, 16.0F), new Vector3f(8.0F, 8.0F, 16.0F), face, block.getConnectedTextures.getTextureForCorner(6, connections), EnumFacing.SOUTH, rot, null, scale, true))
        list.add(faceBakery.makeBakedQuad(new Vector3f(8.0F, 0.0F, 16.0F), new Vector3f(16.0F, 8.0F, 16.0F), face, block.getConnectedTextures.getTextureForCorner(7, connections), EnumFacing.SOUTH, rot, null, scale, true))
        list.add(faceBakery.makeBakedQuad(new Vector3f(0.0F, 8.0F, 16.0F), new Vector3f(8.0F, 16.0F, 16.0F), face, block.getConnectedTextures.getTextureForCorner(4, connections), EnumFacing.SOUTH, rot, null, scale, true))
        list.add(faceBakery.makeBakedQuad(new Vector3f(8.0F, 8.0F, 16.0F), new Vector3f(16.0F, 16.0F, 16.0F), face, block.getConnectedTextures.getTextureForCorner(5, connections), EnumFacing.SOUTH, rot, null, scale, true))

        if (block != null && block.isTranslucent && world != null) {
            list.add(faceBakery.makeBakedQuad(new Vector3f(0.0F, 0.0F, 16.0F), new Vector3f(8.0F, 8.0F, 15.999F), face, block.getConnectedTextures.getTextureForCorner(6, connections), EnumFacing.NORTH, rot, null, scale, true))
            list.add(faceBakery.makeBakedQuad(new Vector3f(8.0F, 0.0F, 16.0F), new Vector3f(16.0F, 8.0F, 15.999F), face, block.getConnectedTextures.getTextureForCorner(7, connections), EnumFacing.NORTH, rot, null, scale, true))
            list.add(faceBakery.makeBakedQuad(new Vector3f(0.0F, 8.0F, 16.0F), new Vector3f(8.0F, 16.0F, 15.999F), face, block.getConnectedTextures.getTextureForCorner(4, connections), EnumFacing.NORTH, rot, null, scale, true))
            list.add(faceBakery.makeBakedQuad(new Vector3f(8.0F, 8.0F, 16.0F), new Vector3f(16.0F, 16.0F, 15.999F), face, block.getConnectedTextures.getTextureForCorner(5, connections), EnumFacing.NORTH, rot, null, scale, true))
        }
    }

    def lookUpRotationForFace(face : EnumFacing) : ModelRotation = {
        face match {
            case EnumFacing.UP =>
                ModelRotation.X90_Y0
            case EnumFacing.DOWN =>
                ModelRotation.X270_Y0
            case EnumFacing.NORTH =>
                ModelRotation.X0_Y180
            case EnumFacing.EAST =>
                ModelRotation.X0_Y270
            case EnumFacing.SOUTH =>
                ModelRotation.X0_Y0
            case EnumFacing.WEST =>
                ModelRotation.X0_Y90
            case _ =>
                ModelRotation.X0_Y0
        }
    }

    override def getGeneralQuads: util.List[BakedQuad] = new util.ArrayList[BakedQuad]()
    override def getParticleTexture: TextureAtlasSprite = block.getConnectedTextures.corners
    override def isBuiltInRenderer: Boolean = false
    override def getItemCameraTransforms: ItemCameraTransforms = {
        val tRSRTransformation = TRSRTransformation.blockCenterToCorner(new TRSRTransformation(
            new javax.vecmath.Vector3f(-0.325F, -0.2F, -0.5F),
            TRSRTransformation.quatFromYXZDegrees(new javax.vecmath.Vector3f(0F, 0.25F, 0F)),
            new javax.vecmath.Vector3f(0.375F, 0.375F, 0.375F),
            null))
        new ItemCameraTransforms( tRSRTransformation.toItemTransform, ItemTransformVec3f.DEFAULT,
            ItemTransformVec3f.DEFAULT, ItemTransformVec3f.DEFAULT, ItemTransformVec3f.DEFAULT, ItemTransformVec3f.DEFAULT)
    }
    override def isAmbientOcclusion: Boolean = true
    override def isGui3d: Boolean = true
}
